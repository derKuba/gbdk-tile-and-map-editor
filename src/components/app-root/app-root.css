header {
  color: white;
  height: 30px;
  display: flex;
  align-items: center;
  box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.26);
}

h1 {
  font-size: 1rem;
  font-weight: 500;
  color: #fff;
  padding: 0 12px;
}


.gb-green {
  background-color: #6b790e;
}

.gb-grey {
  background-color: #c5bfbf;
}

.gb-dark-grey {
  background-color: #5a5760;
}

.gb-purple {
  background-color: #a52059;
}


.disclaimer {
  position: absolute;
  bottom: 0;
  right: 0;
  font-size: 12px;
}