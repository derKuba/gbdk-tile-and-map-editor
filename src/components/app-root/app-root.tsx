/* eslint-disable max-lines, max-lines-per-function */

import { Component, h, State } from '@stencil/core';

import { Settings } from '../../models/settingsModel';
import { PaintModus } from '../../models/paintModus';

import Tunnel from '../../provider/StateProvider';

@Component({
    tag: 'app-root',
    styleUrl: 'app-root.css',
    shadow: true,
})
export class AppRoot {
    @State() indexTileMatrix: number = 0;
    @State() matrix: number[][][] = [];
    @State() choosenColour: number = 2;
    @State() isModalOpen: boolean = false;
    @State() copyIndex: number = 0;
    @State() paintModus: PaintModus = PaintModus.BRUSH;

    @State() settings: Settings = {
        fieldHeight: 8,
        fieldWidth: 8,
        spriteAmount: 12,
        isGridVisible: true,
        isFieldNumberVisible: false,
    };

    private updateColor = color => {
        this.choosenColour = color;
    };

    private changeTileMatrix = index => {
        this.indexTileMatrix = index;
    };

    private setSettings = ({ fieldHeight, fieldWidth, spriteAmount, isGridVisible, isFieldNumberVisible }) => {
        this.settings = {
            fieldHeight,
            fieldWidth,
            spriteAmount,
            isGridVisible,
            isFieldNumberVisible,
        };
        this.matrix = this.createMatrixField(spriteAmount, fieldHeight, fieldWidth);
    };

    private setMatrix = matrix => {
        this.matrix = matrix;
    };

    private setPaintModus = modus => {
        this.paintModus = modus;
    };

    private createMatrixField = (spriteAmount, fieldHeight, fieldWidth) => {
        const matrix: number[][][] = [];
        for (let i = 0; i < spriteAmount; i++) {
            matrix.push(Array.from({ length: fieldHeight }, () => Array.from({ length: fieldWidth }, () => 0)));
        }

        return matrix;
    };

    private setIsModalOpen = modalState => {
        this.isModalOpen = modalState;
    };

    private pasteMatrix = () => {
        this.matrix[this.indexTileMatrix] = this.matrix[this.copyIndex];
        this.matrix = [...this.matrix];
    };

    private setCopyIndex = () => {
        this.copyIndex = this.indexTileMatrix;
    };

    componentWillLoad() {
        const { spriteAmount, fieldHeight, fieldWidth } = this.settings;
        this.matrix = this.createMatrixField(spriteAmount, fieldHeight, fieldWidth);

        /*
                this.matrix[0] = [
                    [1, 1, 1, 0, 1, 1, 1, 0],
                    [0, 0, 1, 0, 1, 0, 0, 1],
                    [0, 0, 1, 0, 1, 1, 1, 0],
                    [0, 0, 1, 0, 1, 0, 0, 0],
                    [0, 1, 1, 0, 1, 0, 0, 0],
                    [1, 0, 1, 0, 1, 0, 0, 0],
                    [0, 1, 0, 0, 1, 0, 0, 0],
                    [0, 0, 0, 0, 0, 0, 0, 0],
                ];*/
    }

    render() {
        const tunnelState = {
            matrix: this.matrix,
            indexTileMatrix: this.indexTileMatrix,
            choosenColour: this.choosenColour,
            updateColor: this.updateColor,
            changeTileMatrix: this.changeTileMatrix,
            setSettings: this.setSettings,
            settings: this.settings,
            setMatrix: this.setMatrix,
            paintModus: this.paintModus,
            setPaintModus: this.setPaintModus,
            isModalOpen: this.isModalOpen,
            setIsModalOpen: this.setIsModalOpen,
            copyIndex: this.copyIndex,
            setCopyIndex: this.setCopyIndex,
            pasteMatrix: this.pasteMatrix,
        };

        return (
            <div>
                <header class="gb-purple">
                    <h1>GB Sprite Tile Creator</h1>
                </header>

                <main>
                    <Tunnel.Provider state={tunnelState}>
                        <stencil-router>
                            <stencil-route-switch scrollTopOffset={0}>
                                <stencil-route url="/" component="app-home" exact={true} />
                                <stencil-route url="/docs" component="app-home" />
                                <stencil-route url="/profile/:name" component="app-profile" />
                            </stencil-route-switch>
                        </stencil-router>
                    </Tunnel.Provider>
                </main>

                <div class="disclaimer">
                    Icons made by{' '}
                    <a href="https://www.flaticon.com/authors/freepik" title="Freepik">
                        Freepik
                    </a>{' '}
                    from{' '}
                    <a href="https://www.flaticon.com/" title="Flaticon">
                        {' '}
                        www.flaticon.com
                    </a>
                </div>
            </div>
        );
    }
}
