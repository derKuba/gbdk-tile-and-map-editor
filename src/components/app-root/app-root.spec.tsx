import { newSpecPage } from '@stencil/core/dist/testing';
import { h } from '@stencil/core';

import { AppRoot } from './app-root';

describe('app-root', () => {
    it('builds', () => {
        expect(new AppRoot()).toBeTruthy();
    });
    it('matches snapshot', async () => {
        const page = await newSpecPage({
            components: [AppRoot],
            template: () => <app-root />,
        });

        expect(page.root).toMatchSnapshot();
    });

    it('should update choosenColour', async () => {
        const rootApp = new AppRoot();
        rootApp.updateColor(23);
        expect(rootApp.choosenColour).toBe(23);
    });

    it('should update matrix', async () => {
        const rootApp = new AppRoot();
        rootApp.setMatrix([]);
        expect(rootApp.matrix).toEqual([]);
    });

    it('should update modal', async () => {
        const rootApp = new AppRoot();
        rootApp.setIsModalOpen(true);
        expect(rootApp.isModalOpen).toBe(true);
    });

    it('should update copyIndex and tilematrix', async () => {
        const rootApp = new AppRoot();
        rootApp.changeTileMatrix(23);
        rootApp.setCopyIndex();
        expect(rootApp.copyIndex).toBe(23);
        expect(rootApp.indexTileMatrix).toBe(23);
    });

    it('should update settings', async () => {
        const rootApp = new AppRoot();
        rootApp.setSettings({
            fieldHeight: 100,
            fieldWidth: 200,
            spriteAmount: 23,
            isGridVisible: false,
            isFieldNumberVisible: false,
        });
        expect(rootApp.settings).toEqual({
            fieldHeight: 100,
            fieldWidth: 200,
            isFieldNumberVisible: false,
            isGridVisible: false,
            spriteAmount: 23,
        });
    });

    it('should update paintModus', async () => {
        const rootApp = new AppRoot();
        rootApp.setPaintModus(1);
        expect(rootApp.paintModus).toBe(1);
    });

    it('should pasteMatrix', async () => {
        const rootApp = new AppRoot();
        rootApp.matrix = rootApp.createMatrixField(2, 8, 8);
        rootApp.matrix[0] = [
            [1, 1, 1, 0, 1, 1, 1, 0],
            [0, 0, 1, 0, 1, 0, 0, 1],
            [0, 0, 1, 0, 1, 1, 1, 0],
            [0, 0, 1, 0, 1, 0, 0, 0],
            [0, 1, 1, 0, 1, 0, 0, 0],
            [1, 0, 1, 0, 1, 0, 0, 0],
            [0, 1, 0, 0, 1, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
        ];

        rootApp.changeTileMatrix(0);
        rootApp.setCopyIndex();

        rootApp.changeTileMatrix(1);
        rootApp.pasteMatrix(1);
        expect(rootApp.matrix).toEqual([
            [
                [1, 1, 1, 0, 1, 1, 1, 0],
                [0, 0, 1, 0, 1, 0, 0, 1],
                [0, 0, 1, 0, 1, 1, 1, 0],
                [0, 0, 1, 0, 1, 0, 0, 0],
                [0, 1, 1, 0, 1, 0, 0, 0],
                [1, 0, 1, 0, 1, 0, 0, 0],
                [0, 1, 0, 0, 1, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0],
            ],
            [
                [1, 1, 1, 0, 1, 1, 1, 0],
                [0, 0, 1, 0, 1, 0, 0, 1],
                [0, 0, 1, 0, 1, 1, 1, 0],
                [0, 0, 1, 0, 1, 0, 0, 0],
                [0, 1, 1, 0, 1, 0, 0, 0],
                [1, 0, 1, 0, 1, 0, 0, 0],
                [0, 1, 0, 0, 1, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0],
            ],
        ]);
    });
});
