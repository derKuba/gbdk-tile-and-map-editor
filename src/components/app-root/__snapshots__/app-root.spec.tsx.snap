// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-root matches snapshot 1`] = `
<app-root>
  <mock:shadow-root>
    <div>
      <header class="gb-purple">
        <h1>
          GB Sprite Tile Creator
        </h1>
      </header>
      <main>
        <stencil-router>
          <stencil-route-switch scrolltopoffset="0">
            <stencil-route component="app-home" exact="" url="/"></stencil-route>
            <stencil-route component="app-profile" url="/profile/:name"></stencil-route>
          </stencil-route-switch>
        </stencil-router>
      </main>
      <div class="disclaimer">
        Icons made by
        <a href="https://www.flaticon.com/authors/freepik" title="Freepik">
          Freepik
        </a>
        from
        <a href="https://www.flaticon.com/" title="Flaticon">
          www.flaticon.com
        </a>
      </div>
    </div>
  </mock:shadow-root>
</app-root>
`;
